### PR TITLE
[iter] Add .transform() function for arbitrary generator transforms

### PR DESCRIPTION
**Pull request by ericl**
## Why are these changes needed?

You can already implement this today with a stateful callable class, but it can be more intuitive to write a function instead.

@michaelzhiluo you might want to try this out for your MAML implementation.

### CODE REVIEW

File: src/ray/object_manager/plasma/store.cc [hunk 1]
```diff
@@ -730,7 +730,7 @@ void PlasmaStore::DisconnectClient(int client_fd) {
   loop_->RemoveFileEvent(client_fd);
   // Close the socket.
   close(client_fd);
-  ARROW_LOG(INFO) << "Disconnecting client on fd " << client_fd;
+  ARROW_LOG(DEBUG) << "Disconnecting client on fd " << client_fd;
```
Comment (wuisawesome) [inline src/ray/object_manager/plasma/store.cc:733]: This probably shouldn't be here

File: python/ray/util/iter.py [hunk 2]
```diff
@@ -193,10 +193,36 @@ def _with_transform(self, local_it_fn, name):
...
+    def transform(self, fn: Callable[[Iterable[T]], Iterable[U]]
+                  ) -> "ParallelIterator[U]":
+        """Remotely transform the iterator.
+
+        This is advanced version of for_each that allows you to apply arbitrary
+        generator transformations over the iterator. Prefer to use .for_each()
+        when possible for simplicity.
+
+        Args:
+            fn (func): function to use to transform the iterator. The function
```
Comment (wuisawesome) [inline python/ray/util/iter.py:205]: Can we specify that `fn` is called once per worker since this is stateful?

**Review by wuisawesome — CHANGES_REQUESTED** [review]
Mostly lgtm other than the arrow log.

File: python/ray/util/iter.py (hunk 2 above)
Comment (ericl) [reply python/ray/util/iter.py:205]: Done.

File: src/ray/object_manager/plasma/store.cc (hunk 1 above)
Comment (ericl) [reply src/ray/object_manager/plasma/store.cc:733]: It's a nice incidental fix.

**Review by ericl — COMMENTED** [review]
PTAL

**Review by wuisawesome — APPROVED** [review]

**Pull request merged**